sql/core/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala

```diff
@@ -456,11 +456,23 @@ final class DataFrameNaFunctions private[sql](df: DataFrame) {
     val keyExpr = df.col(col.name).expr
     def buildExpr(v: Any) = Cast(Literal(v), keyExpr.dataType)
     val branches = replacementMap.flatMap { case (source, target) =>
-      Seq(buildExpr(source), buildExpr(target))
+      if (isNaN(source) || isNaN(target)) {
+        col.dataType match {
+          case IntegerType | LongType | ShortType | ByteType => Seq.empty
```

Comment:
Thank you for making a PR and I fully understand this issue, @johnhany97 .

One concern is that the current behavior is consistent with Spark's `CAST` operation which converts `NaN` to `0` during DOUBLE-to-INT casting. Theoretically, while Apache Spark casts the given value to the given column type first, `NaN` becomes `0`.
```scala
scala> Seq(Double.NaN, 0.0).toDF.selectExpr("cast(value as int)").show
+-----+
|value|
+-----+
|    0|
|    0|
+-----+
```
So, it's a natural behavior in that sequence. However, I agree that `na` function specially needs this fix.

Hi, @gatorsmile and @cloud-fan . How do you think about this PR?

Reply:
checked with scala
```
scala> Float.NaN == 0
res0: Boolean = false

scala> Float.NaN.toInt == 0
res1: Boolean = true
```

This is also true in Spark. When comparing float and int, we cast int to float to compare, so `NaN != 0`.

I think it's a bug that we cast the value to the column type and compare. We shouldn't do any cast and let the type coercion rules to do proper cast for `CaseKeyWhen`

Reply:
Thank you for your guide, @cloud-fan !